header-footer-grid/Core/Magic_Tags.php

```diff
@@ -0,0 +1,410 @@
+<?php
+/**
+ * Handles Magic Tags
+ *
+ * @package Neve
+ */
+
+namespace HFG\Core;
+
+/**
+ * Class Short_Codes
+ *
+ * @package Neve\Views\Pluggable
+ */
+class Magic_Tags {
+	/**
+	 * The magic tags options used for Customizer.
+	 *
+	 * @var array
+	 */
+	private $options = [];
+
+	/**
+	 * Available short codes.
+	 *
+	 * @var array
+	 */
+	private $magic_tags = [];
+
+	/**
+	 * Magic tag regex.
+	 *
+	 * @var string
+	 */
+	static private $magic_tag_regex;
+
+	/**
+	 * Holds the instance of the class.
+	 *
+	 * @var Magic_Tags
+	 */
+	static private $_instance;
+
+	/**
+	 * Returns the instance of the class.
+	 *
+	 * @return Magic_Tags
+	 */
+	public static function get_instance() {
+		if ( null === self::$_instance ) {
+			self::$_instance = new self();
+		}
+
+		return self::$_instance;
+	}
+
+	/**
+	 * Function that is run after instantiation.
+	 *
+	 * @return void
+	 */
+	public function __construct() {
+		$this->setup_config();
+		if ( class_exists( 'WooCommerce', false ) ) {
+			$this->magic_tags = array_merge(
+				$this->magic_tags,
+				[
+					'product_price',
+					'product_title',
+					'cart_link',
+					'checkout_link',
+				]
+			);
+		}
+		self::$magic_tag_regex = implode( '|', $this->magic_tags );
+	}
+
+	/**
+	 * Do the magic tags inside the string.
+	 *
+	 * @param string $input the input.
+	 *
+	 * @return string|
+	 */
+	public function do_magic_tags( $input ) {
+		if ( empty( self::$magic_tag_regex ) ) {
+			return $input;
+		}
+
+		if ( ! preg_match( '/\\{\\w+\\}/', $input ) ) {
+			return $input;
+		}
+
+		return preg_replace_callback(
+			'/\\{\b(?:' . self::$magic_tag_regex . ')\b\\}/',
+			[
+				$this,
+				'do_magic_tag',
+			],
+			$input
+		);
+	}
+
+	/**
+	 * Do single magic tag.
+	 *
+	 * @param string $tag the magic tag.
+	 *
+	 * @return string
+	 */
+	private function do_magic_tag( $tag ) {
+		if ( is_array( $tag ) ) {
+			$tag = reset( $tag );
+		}
+
+		$tag = substr( $tag, 1, - 1 );
```

Comment:
@abaicus I think we can do a trim here and also check if the method exists to be on the safe side?

Reply:
@selul Sure, it makes sense! 👍